model/src/main/kotlin/DependencyGraph.kt

```diff
@@ -0,0 +1,240 @@
+/*
+ * Copyright (C) 2021 Bosch.IO GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.model
+
+import com.fasterxml.jackson.annotation.JsonAlias
+import com.fasterxml.jackson.annotation.JsonInclude
+
+import java.lang.IllegalStateException
+import java.util.SortedSet
+
+/**
+ * A data class that represents the graph of dependencies of a project.
+ *
+ * This class holds information about a project's scopes and their dependencies in a format that minimizes the
+ * consumption of memory. In projects with many scopes there is often a high degree of duplication in the
+ * dependencies of the single scopes. To avoid this, this class aims to share as many parts of the dependency graph as
+ * possible between the different scopes. Ideally, there is only a single dependency graph containing the dependencies
+ * used by all scopes. This is not always possibles due to inconsistencies in dependency relations. (A package uses
+ * different dependencies in different scopes.) Then the dependency graph is split into multiple fragments, and each
```

Comment:
Is what you call a "fragment" simply a "sub-tree"? If so, does that term maybe make more sense as it is more telling?

Reply:
It is hard to find a precise and correct wording here, and I was certainly not always consistent with terms like 'graph', 'tree', or 'fragment'. 'Sub-tree' is IMHO not fully correct because the structure is more complex than a typical tree: There can be multiple edges leading to a node, so there is not necessarily a unique path from a node to the root.

The dependency graph can naturally consist of multiple fragments. What I wanted to express in this comment was that in each fragment a package occurs only once with a specific set of dependencies. If the package is also used with different dependencies, it must be added to a new fragment.

Reply:
Some fields of `DependencyGraph` have been renamed to make their intension clearer.